### ISSUE TITLE
Webdav COPY must use TUS for its internal uploads

### ISSUE THREAD

**Issue by PVince81** [owncloud/ocis-reva#173]
Same steps as https://github.com/owncloud/ocis-reva/issues/171

Except specify "Depth: infinity" as header.

See the log output in https://github.com/owncloud/ocis-reva/issues/171 which shows that uploading the files inside the recursion is [sending a PUT request](https://github.com/cs3org/reva/blob/master/internal/http/services/owncloud/ocdav/copy.go#L260) but TUS is answering on the other side, expecting TUS headers, etc.

It seems that put.go already contains a lot of logic for using the TUS client, we should refactor this in a common location to be able to reuse it anywhere in the ocdav service.

@butonic FYI

**Comment by PVince81** [comment]
I can try fixing this but will need some guidance about creating a common location / best practices

**Comment by PVince81** [comment]
Ideally https://github.com/cs3org/reva/pull/714 should be merged first

**Comment by PVince81** [comment]
After discussing with @butonic:

- ideal would be a COPY operation on the storage, but currently not supported on CS3 API
- as we want a quick solution now, we will copy the TUS upload code from put.go into copy.go to make it work

**Comment by butonic** [comment]
the dataprovider is actually out of scope of CS3. hm using tus.io we could use the `concatenation` extension and the [`upload-concat` header](https://tus.io/protocols/resumable-upload.html#upload-concat) to cover cross storage assembly using tus.

The upload parts can be absolute URLs, so we can let the receiving data provider fetch the file. The ocdav handler would call InitiateDownload on the storagprovider and use that URL in a TUS POST with `Upload-Concat: final; https://dataprovider.tld/path/to/file`. Something to discuss with @labkode because it makes more sense to be able to specify final and a list of upload part urls in the CS3 InitiateUpload request as well.

**Comment by PVince81** [comment]
As far as I understand upload-concat is also there only for chunking. So if we want to chunk-upload to another backend storage, then yes.

For the first implementation I'd like to stick to a single request for that one upload. My goal is to fix the regression and [make litmus tests pass](https://github.com/owncloud/ocis-reva/issues/18), not yet improve performance.

**Comment by PVince81** [comment]
Fix is here: https://github.com/cs3org/reva/pull/717
